### Capstone Project
#### Data Wrangling

Using 'RAW_interactions.csv' & 'RAW_recipes.csv') from:
https://www.kaggle.com/datasets/shuyangli94/food-com-recipes-and-user-interactions

#### Data Collection

* Goal: Organize your data to streamline the next steps of your
capstone
    - Data loading
    - Data joining
    - Hint: Data Collection will require the use of the pandas library,
and functions like read_csv(), depending on the type of data
you want to read in!
    - Hint: when adding one dataset to another, make sure you use
the right function: you might want to merge, join, or
concatenate

In [56]:
# import libraries
import pandas as pd
import matplotlib.pyplot as plt

#### Load Data

In [2]:
interactions = pd.read_csv('data/RAW_interactions.csv')
recipes = pd.read_csv('data/RAW_recipes.csv')

#### Data Definition

○ Goal: Gain an understanding of your data features to inform the
next steps of your project.
- Column names
- Data types
- Description of the columns
- Counts and percents unique values
- Ranges of values

* Do your column names correspond to what those columns
store?
* Check the data types of your columns. Are they sensible?
* Calculate summary statistics for each of your columns, such as mean, median, mode, standard deviation, range, and number of unique values. What does this tell you about your data? What do you now need to investigate?


#### Data Cleaning
○ Goal: Clean up the data in order to prepare it for the next steps of
your project.
- NA or missing values
- Duplicates


### Recipes Dataset

In [3]:
recipes.columns

Index(['name', 'id', 'minutes', 'contributor_id', 'submitted', 'tags',
       'nutrition', 'n_steps', 'steps', 'description', 'ingredients',
       'n_ingredients'],
      dtype='object')

In [4]:
recipes.index

RangeIndex(start=0, stop=231637, step=1)

In [5]:
recipes.dtypes

name              object
id                 int64
minutes            int64
contributor_id     int64
submitted         object
tags              object
nutrition         object
n_steps            int64
steps             object
description       object
ingredients       object
n_ingredients      int64
dtype: object

In [6]:
recipes.head()

name      id  minutes  \
0  arriba   baked winter squash mexican style  137739       55   
1            a bit different  breakfast pizza   31490       30   
2                   all in the kitchen  chili  112140      130   
3                          alouette  potatoes   59389       45   
4          amish  tomato ketchup  for canning   44061      190   

   contributor_id   submitted  \
0           47892  2005-09-16   
1           26278  2002-06-17   
2          196586  2005-02-25   
3           68585  2003-04-14   
4           41706  2002-10-25   

                                                tags  \
0  ['60-minutes-or-less', 'time-to-make', 'course...   
1  ['30-minutes-or-less', 'time-to-make', 'course...   
2  ['time-to-make', 'course', 'preparation', 'mai...   
3  ['60-minutes-or-less', 'time-to-make', 'course...   
4  ['weeknight', 'time-to-make', 'course', 'main-...   

                                    nutrition  n_steps  \
0       [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1   [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3   [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4   [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   

                                               steps  \
0  ['make a choice and proceed with recipe', 'dep...   
1  ['preheat oven to 425 degrees f', 'press dough...   
2  ['brown ground beef in large pot', 'add choppe...   
3  ['place potatoes in a large pot of lightly sal...   
4  ['mix all ingredients& boil for 2 1 / 2 hours ...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   

                                         ingredients  n_ingredients  
0  ['winter squash', 'mexican seasoning', 'mixed ...              7  
1  ['prepared pizza crust', 'sausage patty', 'egg...              6  
2  ['ground beef', 'yellow onions', 'diced tomato...             13  
3  ['spreadable cheese with garlic and herbs', 'n...             11  
4  ['tomato juice', 'apple cider vinegar', 'sugar...              8

In [7]:
recipes.isna().sum()

name                 1
id                   0
minutes              0
contributor_id       0
submitted            0
tags                 0
nutrition            0
n_steps              0
steps                0
description       4979
ingredients          0
n_ingredients        0
dtype: int64

#### Drop NaN
Drop NaN in names. Do not know if 'description' will be a key part of the model, so I will keep records with empty descriptions for now. Will drop the 1 record with an emty name

In [8]:
recipes = recipes[recipes['name'].notna()]

In [9]:
# rename some columns
# df.rename(columns={'oldName1': 'newName1', 'oldName2': 'newName2'}, inplace=True)
recipes = recipes.rename(columns = {'contributor_id':'user_id','name':'recipe','id':'recipe_id'})
recipes.head()

recipe  recipe_id  minutes  user_id  \
0  arriba   baked winter squash mexican style     137739       55    47892   
1            a bit different  breakfast pizza      31490       30    26278   
2                   all in the kitchen  chili     112140      130   196586   
3                          alouette  potatoes      59389       45    68585   
4          amish  tomato ketchup  for canning      44061      190    41706   

    submitted                                               tags  \
0  2005-09-16  ['60-minutes-or-less', 'time-to-make', 'course...   
1  2002-06-17  ['30-minutes-or-less', 'time-to-make', 'course...   
2  2005-02-25  ['time-to-make', 'course', 'preparation', 'mai...   
3  2003-04-14  ['60-minutes-or-less', 'time-to-make', 'course...   
4  2002-10-25  ['weeknight', 'time-to-make', 'course', 'main-...   

                                    nutrition  n_steps  \
0       [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1   [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3   [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4   [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   

                                               steps  \
0  ['make a choice and proceed with recipe', 'dep...   
1  ['preheat oven to 425 degrees f', 'press dough...   
2  ['brown ground beef in large pot', 'add choppe...   
3  ['place potatoes in a large pot of lightly sal...   
4  ['mix all ingredients& boil for 2 1 / 2 hours ...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   

                                         ingredients  n_ingredients  
0  ['winter squash', 'mexican seasoning', 'mixed ...              7  
1  ['prepared pizza crust', 'sausage patty', 'egg...              6  
2  ['ground beef', 'yellow onions', 'diced tomato...             13  
3  ['spreadable cheese with garlic and herbs', 'n...             11  
4  ['tomato juice', 'apple cider vinegar', 'sugar...              8

In [10]:
recipes.select_dtypes('object')

recipe   submitted  \
0         arriba   baked winter squash mexican style  2005-09-16   
1                   a bit different  breakfast pizza  2002-06-17   
2                          all in the kitchen  chili  2005-02-25   
3                                 alouette  potatoes  2003-04-14   
4                 amish  tomato ketchup  for canning  2002-10-25   
...                                              ...         ...   
231632                                   zydeco soup  2012-08-29   
231633                              zydeco spice mix  2013-01-09   
231634                     zydeco ya ya deviled eggs  2008-06-07   
231635        cookies by design   cookies on a stick  2008-04-15   
231636  cookies by design   sugar shortbread cookies  2008-04-15   

                                                     tags  \
0       ['60-minutes-or-less', 'time-to-make', 'course...   
1       ['30-minutes-or-less', 'time-to-make', 'course...   
2       ['time-to-make', 'course', 'preparation', 'mai...   
3       ['60-minutes-or-less', 'time-to-make', 'course...   
4       ['weeknight', 'time-to-make', 'course', 'main-...   
...                                                   ...   
231632  ['ham', '60-minutes-or-less', 'time-to-make', ...   
231633  ['15-minutes-or-less', 'time-to-make', 'course...   
231634  ['60-minutes-or-less', 'time-to-make', 'course...   
231635  ['30-minutes-or-less', 'time-to-make', 'course...   
231636  ['30-minutes-or-less', 'time-to-make', 'course...   

                                          nutrition  \
0             [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]   
1         [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]   
2        [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]   
3         [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]   
4         [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]   
...                                             ...   
231632  [415.2, 26.0, 34.0, 26.0, 44.0, 21.0, 15.0]   
231633        [14.8, 0.0, 2.0, 58.0, 1.0, 0.0, 1.0]   
231634         [59.2, 6.0, 2.0, 3.0, 6.0, 5.0, 0.0]   
231635    [188.0, 11.0, 57.0, 11.0, 7.0, 21.0, 9.0]   
231636     [174.9, 14.0, 33.0, 4.0, 4.0, 11.0, 6.0]   

                                                    steps  \
0       ['make a choice and proceed with recipe', 'dep...   
1       ['preheat oven to 425 degrees f', 'press dough...   
2       ['brown ground beef in large pot', 'add choppe...   
3       ['place potatoes in a large pot of lightly sal...   
4       ['mix all ingredients& boil for 2 1 / 2 hours ...   
...                                                   ...   
231632  ['heat oil in a 4-quart dutch oven', 'add cele...   
231633        ['mix all ingredients together thoroughly']   
231634  ['in a bowl , combine the mashed yolks and may...   
231635  ['place melted butter in a large mixing bowl a...   
231636  ['whip sugar and shortening in a large bowl , ...   

                                              description  \
0       autumn is my favorite time of year to cook! th...   
1       this recipe calls for the crust to be prebaked...   
2       this modified version of 'mom's' chili was a h...   
3       this is a super easy, great tasting, make ahea...   
4       my dh's amish mother raised him on this recipe...   
...                                                   ...   
231632  this is a delicious soup that i originally fou...   
231633    this spice mix will make your taste buds dance!   
231634                          deviled eggs, cajun-style   
231635  i've heard of the 'cookies by design' company,...   
231636  i've heard of the 'cookies by design' company,...   

                                              ingredients  
0       ['winter squash', 'mexican seasoning', 'mixed ...  
1       ['prepared pizza crust', 'sausage patty', 'egg...  
2       ['ground beef', 'yellow onions', 'diced tomato...  
3       ['spreadable cheese with garlic and herbs', 'n...  
4       ['tomato juice', 'apple cider vinegar', 'sugar...  
...      

In [11]:
# Change submitted to datetime
recipes['submitted'] = pd.to_datetime(recipes['submitted'])
# recipes['nutrition'] this is a string should I change it to an array

In [12]:
# look at first & last dates
(recipes['submitted'].min(),recipes['submitted'].max())

(Timestamp('1999-08-06 00:00:00'), Timestamp('2018-12-04 00:00:00'))

In [13]:
recipes[['submitted','n_steps','n_ingredients','minutes']].describe()

submitted        n_steps  n_ingredients  \
count                         231636  231636.000000  231636.000000   
mean   2006-11-14 01:53:01.122105344       9.765516       9.051149   
min              1999-08-06 00:00:00       0.000000       1.000000   
25%              2004-09-16 00:00:00       6.000000       6.000000   
50%              2007-01-23 00:00:00       9.000000       9.000000   
75%              2008-10-29 00:00:00      12.000000      11.000000   
max              2018-12-04 00:00:00     145.000000      43.000000   
std                              NaN       5.995136       3.734803   

            minutes  
count  2.316360e+05  
mean   9.398587e+03  
min    0.000000e+00  
25%    2.000000e+01  
50%    4.000000e+01  
75%    6.500000e+01  
max    2.147484e+09  
std    4.461973e+06

##### Notes
Will investigate 0 minutes and 0 steps and recipes with only 1 ingredient

In [14]:
# find where n_steps is 0 and/or minutes is 0
zero_minutes = recipes[recipes['minutes']==0]
zero_steps = recipes[recipes['n_steps']==0]

In [15]:
zero_minutes.shape

(1094, 12)

In [16]:
zero_steps

recipe  recipe_id  minutes  user_id  submitted  \
3381  all season bread     176767       90   331268 2006-07-10   

                                                   tags  \
3381  ['time-to-make', 'course', 'main-ingredient', ...   

                                      nutrition  n_steps steps  \
3381  [198.8, 11.0, 70.0, 18.0, 5.0, 5.0, 10.0]        0    []   

                                            description  \
3381  just change the fruit/vegetable in this recipe...   

                                            ingredients  n_ingredients  
3381  ['flour', 'baking soda', 'salt', 'baking powde...             12

In [17]:
# find where n_steps is 0 and/or minutes is 0
one_ingred = recipes[recipes['n_ingredients']==1]
one_ingred.shape

(25, 12)

In [18]:
one_ingred.head(10)

recipe  recipe_id  minutes  user_id  \
6499                    apple cider reduction     521756      155   171084   
13958                   baked corn on the cob     434212       35   542159   
49173                         chocolate  dirt     425260        5   730126   
55087                           clotted cream       9043      900     6258   
74549                 easter hard boiled eggs     354371       12  1119236   
76587                            easy granita     233253        5   226066   
86040                           freezing eggs     153034       35   249074   
103771                healthy salt substitute     134711       10   193499   
106369                        homemade butter        310      395     1551   
106735  homemade panko  japanese bread crumbs     376622        5   157167   

        submitted                                               tags  \
6499   2015-04-19  ['time-to-make', 'course', 'preparation', 'low...   
13958  2010-08-04  ['60-minutes-or-less', 'time-to-make', 'main-i...   
49173  2010-05-13  ['15-minutes-or-less', 'time-to-make', 'course...   
55087  2001-05-19  ['weeknight', 'time-to-make', 'cuisine', 'prep...   
74549  2009-02-06  ['15-minutes-or-less', 'time-to-make', 'course...   
76587  2007-06-07  ['15-minutes-or-less', 'time-to-make', 'main-i...   
86040  2006-01-23  ['60-minutes-or-less', 'time-to-make', 'main-i...   
103771 2005-08-24  ['15-minutes-or-less', 'time-to-make', 'course...   
106369 1999-09-07  ['weeknight', 'time-to-make', 'course', 'main-...   
106735 2009-06-10  ['15-minutes-or-less', 'time-to-make', 'cuisin...   

                                         nutrition  n_steps  \
6499          [14.4, 0.0, 0.0, 8.0, 0.0, 0.0, 1.0]       13   
13958       [113.3, 2.0, 21.0, 0.0, 8.0, 1.0, 8.0]        2   
49173     [201.2, 19.0, 91.0, 0.0, 3.0, 37.0, 8.0]        3   
55087     [410.6, 67.0, 0.0, 1.0, 4.0, 137.0, 1.0]        6   
74549        [68.9, 7.0, 0.0, 2.0, 12.0, 7.0, 0.0]        9   
76587        [55.2, 0.0, 49.0, 0.0, 1.0, 0.0, 4.0]        4   
86040         [11.9, 1.0, 0.0, 0.0, 2.0, 1.0, 0.0]       37   
103771  [230.4, 3.0, 105.0, 48.0, 19.0, 3.0, 14.0]        4   
106369      [68.4, 11.0, 0.0, 0.0, 0.0, 22.0, 0.0]       10   
106735   [199.5, 3.0, 12.0, 21.0, 11.0, 2.0, 12.0]        5   

                                                    steps  \
6499    ['in 6-8 quart stock pan , measure out 2 cups ...   
13958   ['in an oven heated at 350 f , cook corn in hu...   
49173   ['place chips in a coffee grinder', 'grind unt...   
55087   ['cook cream in top of double boiler over simm...   
74549   ['place the eggs in a pot with enough cold sal...   
76587   ['place the grapes or strawberries in the free...   
86040   ['items needed:', 'eggs', 'ziplock bags , vari...   
103771  ['cut up the celery into fairly small pieces',...   
106369  ['place the bowl and blade of a food processor...   
106735  ['carefully trim only the brown crust from fre...   

                                              description  \
6499    i saw a post about this on facebook and gave i...   
13958   one of my neighbors was telling me this is her...   
49173   this chocolate "dirt" is perfect for dusting c...   
55087          this recipe comes close to the real thing.   
74549   a gruszecki tradition and now ours, is that ea...   
76587                       from weight watchers website.   
86040   a tried & true method to have eggs in your hou...   
103771  i need to change this description again.  at f...   
106369                                                NaN   
106735  i was cooking with a japanese friend, and she ...   

                     ingredients  n_ingredients  
6499             ['apple cider']              1  
13958                   ['corn']              1  
49173   ['dark chocolate chips']              1  
55087            ['heavy cream']              1  
74549                    ['egg']              1  
76587    ['red seedless grapes']              1  

##### Notes
Looks like I may have to flag the records with minutes at 0, but there are 1094 recrods so I will keep them for now. There are only 25 records with empty steps and 0 n_steps, so these could potentially will get dropped. The records with only 1 ingrediant look normal.

In [19]:
recipes_clean = recipes[recipes['n_steps']!=0]

In [20]:
recipes_clean.index

Index([     0,      1,      2,      3,      4,      5,      6,      7,      8,
            9,
       ...
       231627, 231628, 231629, 231630, 231631, 231632, 231633, 231634, 231635,
       231636],
      dtype='int64', length=231635)

### Ineractions Dataset

In [21]:
interactions.columns

Index(['user_id', 'recipe_id', 'date', 'rating', 'review'], dtype='object')

In [22]:
interactions.index

RangeIndex(start=0, stop=1132367, step=1)

In [23]:
interactions.head()

user_id  recipe_id        date  rating  \
0    38094      40893  2003-02-17       4   
1  1293707      40893  2011-12-21       5   
2     8937      44394  2002-12-01       4   
3   126440      85009  2010-02-27       5   
4    57222      85009  2011-10-01       5   

                                              review  
0  Great with a salad. Cooked on top of stove for...  
1  So simple, so delicious! Great for chilly fall...  
2  This worked very well and is EASY.  I used not...  
3  I made the Mexican topping and took it to bunk...  
4  Made the cheddar bacon topping, adding a sprin...

In [24]:
interactions.isna().sum()

user_id        0
recipe_id      0
date           0
rating         0
review       169
dtype: int64

In [25]:
# drop interactions with empty reviews
interactions =  interactions[interactions['review'].notna()]

In [26]:
interactions.isna().sum()

user_id      0
recipe_id    0
date         0
rating       0
review       0
dtype: int64

In [27]:
interactions.reset_index(inplace = True)

##### Question
Is there a more efficient way to do this? Can I use inplace in the .notna() function?

In [28]:
# change date to datetime
interactions['date'] = pd.to_datetime(interactions['date'])
interactions.dtypes

index                 int64
user_id               int64
recipe_id             int64
date         datetime64[ns]
rating                int64
review               object
dtype: object

In [29]:
#check min and max dates
(interactions['date'].min(),interactions['date'].max())

(Timestamp('2000-01-25 00:00:00'), Timestamp('2018-12-20 00:00:00'))

In [30]:
interactions['user_id'].nunique()

226419

In [31]:
user_freq = interactions['user_id'].value_counts()
user_freq = pd.DataFrame(user_freq)

In [32]:
reviews_limit = 100;

In [33]:
users = user_freq[user_freq['count'] > reviews_limit]
# type(user_freq)

In [34]:
users.sort_values(by ='count')

count
user_id       
121684     101
1182971    101
1020416    101
804851     101
9580       101
...        ...
128473    3917
169430    4076
383346    4628
37449     5603
424680    7671

[1368 rows x 1 columns]

In [35]:
users.index

Index([ 424680,   37449,  383346,  169430,  128473,   89831,   58104,  133174,
        199848,  305531,
       ...
        153750, 1182971, 1020416,  804851,  498487,  212417,  390058, 1458294,
          9580,  121684],
      dtype='int64', name='user_id', length=1368)

In [48]:
# is_black_or_brown = dogs["color"].isin(["Black","Brown"])
high_interactions = interactions[interactions['user_id'].isin(users.index)]

In [49]:
high_interactions.head()

index  user_id  recipe_id       date  rating  \
3      3   126440      85009 2010-02-27       5   
4      4    57222      85009 2011-10-01       5   
5      5    52282     120345 2005-05-21       4   
6      6   124416     120345 2011-08-06       0   
8      8    76535     134728 2005-09-02       4   

                                              review  
3  I made the Mexican topping and took it to bunk...  
4  Made the cheddar bacon topping, adding a sprin...  
5  very very sweet. after i waited the 2 days i b...  
6  Just an observation, so I will not rate.  I fo...  
8                                         Very good!

In [50]:
high_interactions.drop(columns = 'index')

user_id  recipe_id       date  rating  \
3         126440      85009 2010-02-27       5   
4          57222      85009 2011-10-01       5   
5          52282     120345 2005-05-21       4   
6         124416     120345 2011-08-06       0   
8          76535     134728 2005-09-02       4   
...          ...        ...        ...     ...   
1132184   126435     166739 2007-08-25       5   
1132186   140132      82303 2010-10-01       5   
1132187   199020      82303 2013-03-18       5   
1132188  1122988      82303 2014-07-08       5   
1132196    53932      78003 2009-01-11       4   

                                                    review  
3        I made the Mexican topping and took it to bunk...  
4        Made the cheddar bacon topping, adding a sprin...  
5        very very sweet. after i waited the 2 days i b...  
6        Just an observation, so I will not rate.  I fo...  
8                                               Very good!  
...                                                    ...  
1132184  I was craving my Mom's tea cakes, but don't ha...  
1132186  This is fabulous.  I made it for our dessert t...  
1132187  5 stars for taste! I had a hard time getting m...  
1132188  This was amazingly delicious!  The only change...  
1132196  Very good!  I used regular port as well.  The ...  

[510485 rows x 5 columns]

In [51]:
high_interactions.reset_index(inplace = True)

In [47]:
high_interactions.drop(columns = ['index','level_0'])

user_id  recipe_id       date  rating  \
0        126440      85009 2010-02-27       5   
1         57222      85009 2011-10-01       5   
2         52282     120345 2005-05-21       4   
3        124416     120345 2011-08-06       0   
4         76535     134728 2005-09-02       4   
...         ...        ...        ...     ...   
510480   126435     166739 2007-08-25       5   
510481   140132      82303 2010-10-01       5   
510482   199020      82303 2013-03-18       5   
510483  1122988      82303 2014-07-08       5   
510484    53932      78003 2009-01-11       4   

                                                   review  
0       I made the Mexican topping and took it to bunk...  
1       Made the cheddar bacon topping, adding a sprin...  
2       very very sweet. after i waited the 2 days i b...  
3       Just an observation, so I will not rate.  I fo...  
4                                              Very good!  
...                                                   ...  
510480  I was craving my Mom's tea cakes, but don't ha...  
510481  This is fabulous.  I made it for our dessert t...  
510482  5 stars for taste! I had a hard time getting m...  
510483  This was amazingly delicious!  The only change...  
510484  Very good!  I used regular port as well.  The ...  

[510485 rows x 5 columns]

##### Question

This was a really inefficient way to get my index to make sense. 

In [52]:
interactions[['rating','date']].describe()

rating                           date
count  1.132198e+06                        1132198
mean   4.410971e+00  2009-05-12 16:06:39.593711360
min    0.000000e+00            2000-01-25 00:00:00
25%    4.000000e+00            2007-03-11 00:00:00
50%    5.000000e+00            2008-12-28 00:00:00
75%    5.000000e+00            2011-03-06 00:00:00
max    5.000000e+00            2018-12-20 00:00:00
std    1.264812e+00                            NaN

### Notes
Discard users who have reviewed less than 100 recipes
Is there a limit that wont serve the general public

weak model -- easily fooled by bad or weird data
strong models can also not be fooled by data

applying real world into a model, more modeling the easier it is

end state, have data in table format, numbers that make sense, at that point

csv file, if someone else wanted to use the file in python, a python object, use pkl, a model, etc, some kind of logic, like a .mat


### Useful functions

- loc[] - filter your data by label
- iloc[] - filter your data by indexes
- apply() - execute a function across an axis of a DataFrame
- drop() - drop columns from a DataFrame
- is_unique() - check if a column is a unique identifier
- Series methods, such as str.contains(), which can be used to check if
a certain substring occurs in a string of a Series, and str.extract(),
which can be used to extract capture groups with a certain regex (or
regular expression) pattern
- numPy methods like .where(), to clean columns. Recall that such
methods have the structure: np.where(condition, then, else)
- DataFrame methods to check for null values, such as
df.isnull().values.any()

In [58]:
# Save the data:
# datapath = '../data'
high_interactions.to_csv('interactions_clean.csv')
recipes_clean.to_csv('recipes_clean.csv')